### PR TITLE
fix(route-recognizer): Require leading slashes to optional segments

### PR DESCRIPTION
**Pull request by davismj**
Resolves aurelia/router#575

### CODE REVIEW

**Comment by davismj** [comment]
@EisenbergEffect You might be the only person who would be able to review this code.

@fkleuver Would you also be willing to take a look?

**Comment by fkleuver** [comment]
I don't know enough about the concept of NFA to be able to fully understand these changes just quite yet. I intend to fully understand them one day, but I can't put a timeframe on that though.

One thing that does come to mind when looking at the test cases; wouldn't it be an idea to have a separate test file/folder called "bugs" with specific test cases referring to issues being solved?

Then I'd think a test `router-575` or something, which specifically tests the 'fooBar/:id?' / 'foo/:id?' issue at hand, would really help to preserve information for when you or someone else needs to touch this code in the future. As a "NFA-newbie" I certainly can't see how this test case ensures the issue is solved without really needing to dig in to the specifics of the issue.

Maybe it's obvious to non-newbies though and I'm just nit-picking :)

**Comment by EisenbergEffect** [comment]
I think that's a really good idea. @davismj What do you think? Some way to link from the test back to the GitHub issues would be cool.

**Comment by StrahilKazlachev** [comment]
I think this will require a great deal of discipline.

**Comment by davismj** [comment]
@EisenbergEffect @StrahilKazlachev I'm not opposed to it if there's a good way to do it. 

I'm more interested in getting someone to look at my logic. The previous code was using NFA terminology, like epsilon, but wasn't using it correctly which was the source of the bug. I may have abandoned terminology for this particular algorithm, although I think the logic is the same and implemented properly. Thus, my main concern is that this makes sense to people other than me, and I'm not creating a maintainability nightmare.

**Comment by Alexander-Taran** [comment]
@davismj could you summarize the change?
in a way: 
before it was like this. 
it failed 'here' because of ...
change proposed is..
it affects ..

I'll make take look it through

**Comment by fkleuver** [comment]
@davismj 

I had a more thorough look at it. And you know, it does actually make more sense now.

When I was first trying to understand this logic (before this PR) the usage of epsilon is precisely what had me confused. It made no sense to me to process a state without consuming it. It also seemed to me that it was causing the issue, but I assumed that there was some more complex design at play which I couldn't see.

On the same token, the conditional regex in `segments.js` also had me confused. If your change to it actually fixes the issue without breaking anything else, then I say overall you've succeeded in making this project as a whole **much** more understandable.

The only part I'm still a little unsure about is where you attach the handlers for the current segment to each individual optional state. In the previous logic I believe that could have caused certain false positives (similar to the one in my initial naive PR).
Is this actually solved by removing the epsilon and changing the regex, or would these handlers not have caused false positives even in the previous logic?

**Comment by davismj** [comment]
@fkleuver thanks for the review.

Lets say we have foo/:id?/bar. What the epsilon was trying to do before was connect `foo` to `/bar` and `foo` to `/:id?`. The `/:id?` segment is `/` > `[^/]*`. The previous epsilon was connecting the 'o' in `foo` to the `[^/]*` in id, skipping the `/`. In other words, it was connecting the end of foo to the *end* of `/:id?` instead of the beginning. My code fixes this by connecting to the beginning of `/:id?`. I threw away the epsilon property in the process.
